### PR TITLE
Adds support for asynchronous gauge metric type

### LINKED ISSUE #12641: [Feature Request] [Metrics Framework] Add support of Gauge type of metrics
### Is your feature request related to a problem? Please describe

As part of the https://github.com/opensearch-project/OpenSearch/issues/10238 we have added the Metrics Framework. We need to enhance this by adding the support for Gauge type of metrics. Gauge metrics are required to capture point in time snapshot of values.

### Describe the solution you'd like

Extending it further to add support for Gauge which is a current value at the time it is read. Gauges can be synchronous as well as asynchronous. This issue the scope is to add asynchronous Gauge.

Asynchronous Gauge - When the measurement is exposed via an accessor, use Asynchronous Gauge to invoke the accessor in a callback function

API

```
Closeable createDoubleGauge(String name, String description, String unit, Supplier<Double> valueProvider, Tags tags);
```

### Related component

Other

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Gaganjuneja**
### Description
As part of the https://github.com/opensearch-project/OpenSearch/issues/10238 we have added the Metrics Framework. We need to enhance this by adding the support for Gauge type of metrics. Gauge metrics are required to capture point in time snapshot of values.

### Related Issues
Resolves #12641

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by Gaganjuneja** [comment]
@reta, Please take a look. Thanks!

File: libs/telemetry/src/main/java/org/opensearch/telemetry/metrics/DefaultMetricsRegistry.java [hunk 1]
```diff
@@ -39,6 +43,11 @@ public Histogram createHistogram(String name, String description, String unit) {
         return metricsTelemetry.createHistogram(name, description, unit);
     }
 
+    @Override
+    public Closeable createObservableGauge(String name, String description, String unit, Supplier<Double> valueProvider, Tags tags) {
```
Comment (reta) [inline libs/telemetry/src/main/java/org/opensearch/telemetry/metrics/DefaultMetricsRegistry.java:47]: Thinking about naming, I think we should not use `observable` here - this is just a gauge with value supplier, nothing more:

```suggestion
    public Closeable createGauge(String name, String description, String unit, Supplier<Double> valueProvider, Tags tags) {
```

**Comment by reta** [comment]
> @reta, Please take a look. Thanks!

Looks pretty good, thanks @Gaganjuneja , one minor comment on naming, thank you.

**Review by reta — APPROVED** [review]

**Comment by Gaganjuneja** [comment]
@reta, I have fixed build errors. Please take a look once and merge. Thanks!

**Review by reta — APPROVED** [review]

**Comment by reta** [comment]
> @reta, I have fixed build errors. Please take a look once and merge. Thanks!

@Gaganjuneja please create a documentation issue (if none exists yet), thank you

**Comment by reta** [comment]
@Gaganjuneja could you also please send manual backport to `2.x`, thank you

**Comment by Gaganjuneja** [comment]
> > @reta, I have fixed build errors. Please take a look once and merge. Thanks!
> 
> @Gaganjuneja please create a documentation issue (if none exists yet), thank you

Created one - https://github.com/opensearch-project/documentation-website/issues/6694

**Pull request merged**